### PR TITLE
[jit] `__copy__` for `RecursiveScriptModule`

### LINKED ISSUE #32519: torch::jit::script::Module::clone() is not actually cloning parameters
## 🐛 Bug

If you take a look at the highlighted section: https://github.com/pytorch/pytorch/blob/4314620ba05bc1867f6a63455c4ac77fdfb1018d/torch/csrc/jit/script/module.cpp#L191-L203, you would notice that when `IValue s` is not a module, `Module r` will simply register it. Therefore, the new module will share the same parameters.

## To Reproduce

1. Load any TorchScript model `model`
2. Run `clone = model.clone()`
3. Find any parameter in `model` and that same parameter in `clone`, `isAliasOf` will evaluate to `true`

## Expected behavior

Expect `clone()` to deep copy  everything in a TorchScript model.

## Environment

N/A

## Additional context

N/A

cc @ezyang @gchanan @zou3519 @suo

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32685 [jit] Remove `copy` from public API
* #37538 [jit][api][refactor] remove redundant deepcopy implementation
* **#36830 [jit] `__copy__` for `RecursiveScriptModule`**

Summary:

Test Plan:
build/bin/test_jit

Reviewers:
.

Subscribers:

Tasks:

Tags:

Differential Revision: [D21431012](https://our.internmc.facebook.com/intern/diff/D21431012)

### CODE REVIEW

File: aten/src/ATen/core/ivalue.cpp [hunk 1]
```diff
@@ -470,6 +470,57 @@ std::shared_ptr<ClassType> ivalue::Object::type() const {
...
+IValue IValue::copy() const {
+  IValue copy;
+  switch(tag) {
+    case IValue::Tag::Tensor:
+      copy = IValue(toTensor());
+      break;
+    case IValue::Tag::Tuple: {
+      std::vector<IValue> copied_tuple;
+      for (const auto& e : toTuple()->elements()) {
+        copied_tuple.push_back(e.copy());
```
Comment (zdevito) [inline aten/src/ATen/core/ivalue.cpp:482]: A shallow copy shouldn't be copying the subreferences right?

File: aten/src/ATen/core/ivalue.cpp [hunk 2]
```diff
@@ -470,6 +470,57 @@ std::shared_ptr<ClassType> ivalue::Object::type() const {
...
+      }
+      copy = IValue(ivalue::Tuple::create(copied_tuple));
+    }
+      break;
+    case IValue::Tag::GenericList: {
+      auto list = toList();
+      auto copied_list = c10::impl::GenericList(list.elementType());
+      for (IValue v : list) {
+        copied_list.push_back(v.copy());
```
Comment (zdevito) [inline aten/src/ATen/core/ivalue.cpp:491]: same here

File: aten/src/ATen/core/ivalue.cpp [hunk 3]
```diff
@@ -470,6 +470,57 @@ std::shared_ptr<ClassType> ivalue::Object::type() const {
...
+      }
+      copy = IValue(copied_list);
+    }
+      break;
+    case IValue::Tag::GenericDict: {
+      auto dict = toGenericDict();
+      auto copied_dict = c10::impl::GenericDict(dict.keyType(), dict.valueType());
+      for (const auto& entry : dict) {
+        copied_dict.insert(entry.key().copy(), entry.value().copy());
```
Comment (zdevito) [inline aten/src/ATen/core/ivalue.cpp:500]: and here

**Review by zdevito — COMMENTED** [review]
Is this a shallow copy? It seems to be doing a deep copy without memoization? The behavior of the is function should be aligned with python.

File: aten/src/ATen/core/ivalue.cpp (hunk 1 above)
Comment (jerryzh168) [reply aten/src/ATen/core/ivalue.cpp:482]: I see, will change

File: aten/src/ATen/core/ivalue.cpp [hunk 4]
```diff
@@ -470,6 +470,39 @@ std::shared_ptr<ClassType> ivalue::Object::type() const {
   return type_.type_->expect<ClassType>();
 }
 
+IValue IValue::copy() const {
+  IValue copy;
+  switch(tag) {
+    case IValue::Tag::Tensor:
+      copy = IValue(toTensor());
```
Comment (zdevito) [inline aten/src/ATen/core/ivalue.cpp:477]: These statements are not copying these objects, they are just doing a round trip IValue -> Reference -> IValue

File: torch/csrc/jit/api/object.cpp [hunk 5]
```diff
@@ -35,6 +35,18 @@ void Object::define(const std::string& src, const ResolverPtr& resolver) {
       *type()->name(), src, resolver ? resolver : nativeResolver(), &self);
 }
 
+Object Object::copy() const {
+  Object obj(_ivalue()->compilation_unit(), type());
```
Comment (zdevito) [inline torch/csrc/jit/api/object.cpp:39]: This code is duplicated with `ivalue::Object::copy`, and should just call it instead.

File: torch/csrc/jit/api/object.cpp [hunk 6]
```diff
@@ -35,6 +35,18 @@ void Object::define(const std::string& src, const ResolverPtr& resolver) {
...
+
+  size_t N = type()->numAttributes();
+  for (size_t i = 0; i < N; ++i) {
+    IValue s = _ivalue()->getSlot(i);
+    obj._ivalue()->setAttr(type()->getAttributeName(i), s.copy());
```
Comment (zdevito) [inline torch/csrc/jit/api/object.cpp:44]: calling `s.copy()` makes it a deep copy

**Review by zdevito — COMMENTED** [review]
Thanks for updating this. There are still a number of bugs making it not consistent whether this is a deep or shallow copy. The fact that there are no tests to catch these bugs is concerning as well.

File: aten/src/ATen/core/ivalue.cpp (hunk 4 above)
Comment (jerryzh168) [reply aten/src/ATen/core/ivalue.cpp:477]: Yeah, this is meant to be a shallow copy, so only the IValue wrapper is updated

File: torch/csrc/jit/api/object.cpp (hunk 6 above)
Comment (jerryzh168) [reply torch/csrc/jit/api/object.cpp:44]: I see, so we just need to copy the top level wrapper for shallow copy?

**Review by zdevito — COMMENTED** [review]
The current code is not copying (even shallow) anything. Object and Module are just wrapper classes around `ivalue`. A shallow copy of an object needs to actually create a copy of the underlying ivalue.

**Comment by jerryzh168** [comment]
> The current code is not copying (even shallow) anything. Object and Module are just wrapper classes around `ivalue`. A shallow copy of an object needs to actually create a copy of the underlying ivalue.

so only create a new copy of the underlying ivalue::Object, but don't copy the `c10::IValue` inside the `ivalue::Object`?

**Comment by jerryzh168** [comment]
@zdevito could you review again?

**Review by zdevito — APPROVED** [review]
Thanks!

**Pull request closed without merging**